### ISSUE TITLE
Support dynamic reloading of plugins

### ISSUE THREAD

**Issue by renepickhardt** [ElementsProject/lightning#2325]
Currently every time I change the code of the plugin I need to restart the lightning node which takes some time. It would be cool if there was a way to reload plugins in a running lightning node

**Comment by conscott** [comment]
The correct way to do this is to get inotify events from the filesystem for registered plugins, and I suppose it's an option that should only be compiled in debug mode, HOWEVER, there is a poor man's version of this that will work just fine.

Here is my hacky version:
https://github.com/conscott/lightning/blob/plugin_restarter/reload_plugins.py

I just made a wrapper for lightningd that will kill it and restart it when any file in your plugin dir changes. You can use its like 
```
./reload_plugins.py lightningd/lightningd --plugin-dir ~/c-lightning-plugins/
```
For now I [hardcoded that plugindir](https://github.com/conscott/lightning/blob/plugin_restarter/reload_plugins.py#L41-L42), but with just a bit more work you can make it pull from command line args or from the lightning config file. 

Seems to work okay. If there is more interest in this, I can pull the script out to separate repo and polish it up a bit.

**Comment by conscott** [comment]
I've wanted this for myself, so the issue motivated me enough to make it :)

**Comment by cdecker** [comment]
There are three ways I can see that'd allow a plugin to dynamically reload:

 - `lightningd` detects a change, kills the old instance and starts a new one with the new code
 - the plugin itself manages the reload, detecting its own changes and reloading when something changes (could be part of the framework)
 - we create a `--plugin-runner` argument that is prepended to the `exec` call that is run by `lightningd`, this could then detect changes and handle the killing and restarting of the plugin on its own.

Since this is mostly useful for plugin developers I think I'd vote against the first, in favor of the runner or the self-reloading plugins.

Notice however that there are a few things that cannot be changed when reloading the plugin:

 - Methods, notification subscriptions and hooks that are registered may not be added or removed upon reload (this could be changed and made more dynamic but it'd require support for `manifest`-notifications in `lightningd`)
 - Command line options cannot be changed without restarting `lightningd`, this is in the nature of the command line options themselves, and I don't see a way around this.

**Comment by ZmnSCPxj** [comment]
Closed in #2771
